move placeholder support from curry to partial

kedashoe opened the pull request:
Closes #1363. WIP.
- placeholder support in `partial`
- no more curry1 :emoji:
- don't need [test/internal](https://github.com/ramda/ramda/tree/master/test/internal) anymore, only files in there were curry{2,3}.js and they were only testing R.__ support
- curry about 10% faster according to benchRunner
- added `rearg`. Simple enough function I think, don't really have a strong opinion one way or the other if it should be included.

todo:
- clean up the rest of the tests using `R.__`
- check doc examples for `R.__`
- `partialRight`
- other?

Let me know how it's looking so far.. son is waking up from nap, gotta run! bbl

buzzdecafe commented:
wow, this is a lot of work. thanks so much for taking a whack at this. I am very much looking forward to seeing how this proceeds once your son gets back to sleep

CrossEye commented on src/divide.js:
```diff
@@ -17,9 +17,6 @@ var _curry2 = require('./internal/_curry2');
  *
- *      var half = R.divide(R.__, 2);
- *      half(42); //=> 21
- *
```
I think we'd still want to show how to do this, either with `flip` or with `partial`.

(Unless we ever decide to add things like `divideBy` and cross-reference the documentation.)

kedashoe replied:
This example still lives, I moved it to `partial`. But we can move it back here if that is preferred.

CrossEye commented on src/rearg.js:
```diff
@@ -0,0 +1,45 @@
+ *      var sayHello = R.partial(greet, ['Hello']);
+ *      var sayHelloToMs = R.partial(sayHello, ['Ms.']);
+ *      sayHelloToMs('Jane', 'Jones'); //=> 'Hello, Ms. Jane Jones!'
+ */
```
Are these the examples for another function?  They don't seem to mention `rearg`.

kedashoe replied:
woops yep, I'll update this if we decide to keep rearg

CrossEye commented on test/addIndex.js:
```diff
@@ -68,7 +68,7 @@ describe('addIndex', function() {
       eq(mapFilterIndexed(
         R.multiply,
-        R.gt(R.__, 13),
+        R.lte(13),
```
First off, `R.gt(R.__, x)` is equivalent to `R.lt(x)`, not `R.lte(x)`.  It's an easy mistake to make, as that's how the converses of comparison operators work.  But this is not a converse.

But more importantly, I think if we don't introduce some other function such as `isGt` and simply crossreference them in the docs, then we show how to achieve this by using `gt`, either with `flip` or `partial`.

kedashoe replied:
Sounds good, I'll flip it.

CrossEye commented:
This is great stuff!  I'm sorry I won't be around over the next two weeks to see how it proceeds.

I really, really, like seeing all those `curry1` calls vanishing!  

:rabbit: :tophat:

kedashoe commented:
> wow, this is a lot of work. thanks so much for taking a whack at this. I am very much looking forward to seeing how this proceeds once your son gets back to sleep

won't have much time to code over the weekend, but I'll be sure to keep everyone appraised of what the kids are up to :stuck_out_tongue:

kedashoe commented:
There are several tests like this in various test files

```js
  it('behaves right curried when passed `R.__` for its first argument', function() {
    var lt5 = R.lt(R.__, 5);
    eq(lt5(10), false);
    eq(lt5(5), false);
    eq(lt5(3), true);
});
```

Seems unnecessary to me, shall I just get rid of these? Or use the placeholder with partial for all of them?

buzzdecafe commented:
IMO they should be removed. If you want to test `partial`/`placeholder` behavior that should be in `partial*` tests, or could be consolidated into a placeholder test file. Then if regressions pop up, we can add tests as needed.

kedashoe commented:
Ok I think this can be more thoroughly reviewed now

- updated `partialRight`
- tests for `partial` and `partialRight`
- removed `R.__` from examples
- got rid of `rearg`. can always add it later if there is a clamoring for it

let me know how all that looks and if I've forgot anything

buzzdecafe commented:
I love this, I think it is the right thing to do. I am a bit anxious about regressions this may cause out there; but I would rather err on the side of doing the right thing than legacy support for a lib that is not even at 1.0 yet.

That said, I note that curried functions will throw if the last argument is undefined, but return a reference to the function otherwise:

```js
> R.reduce()();
[Function: f3]
> R.reduce(R.add)();
[Function: f2]
> R.reduce(R.add, 0)();
TypeError: Cannot read property 'reduce' of undefined
```

In my opinion, *all* of these should throw. Invoking a curried function with no arguments makes no sense and should be an error. Others may disagree -- so let's figure it out. And it may be beyond the scope of this PR. 

That said, I am strongly :+1: :cow2: on this PR.

CrossEye commented:
Just back from vacation, and simply triaging issues at the moment.  This one will take more than that.  I'm putting it at the top of my 'to investigate more thoroughly' list.

buzzdecafe commented:
((:bell:))

scott-christopher commented on src/internal/_xreduceBy.js:
```diff
@@ -1,7 +1,6 @@
 var _has = require('./_has');
 var _xfBase = require('./_xfBase');
-
+var curryN = require('./curryN');
```
This should either be `require('../curryN')` or `require('./_curryN')` here.

scott-christopher commented on src/of.js:
```diff
@@ -20,4 +19,4 @@ var _of = require('./internal/_of');
  *      R.of([42]); //=> [[42]]
  */
-module.exports = _curry1(_of);
+module.exports = _of;
```
We might as well just inline `internal/_of` here now.

kedashoe replied:
by inlined do you mean

`module.exports = require('./internal/_of');`

? It doesn't look like our build is equipped to handle that, it leaves the `require` in the output:

`var of = require('./internal/_of');`

Or something else you had in mind?

scott-christopher replied:
Sorry, the suggestions was just to move the contents of `internal/_of` into here and then remove `internal/_of.js`.

kedashoe replied:
Ah! okay, sounds good.

scott-christopher commented on src/reduced.js:
```diff
@@ -20,9 +19,9 @@ var _reduced = require('./internal/_reduced');
  */
 
-module.exports = _curry1(_reduced);
+module.exports = _reduced;
```
And `internal/_reduced` could be inlined here.

scott-christopher commented on src/partialRight.js:
```diff
@@ -26,4 +30,37 @@ var flip = require('./flip');
  *      greetMsJaneJones('Hello'); //=> 'Hello, Ms. Jane Jones!'
  */
-module.exports = _createPartialApplicator(flip(_concat));
+module.exports = _curry2(function partialRight(fn, partialArgs) {
```
Is it perhaps worth simplifying the implementation of `partialRight` by deriving it from `partial`?

Something like:
```js
function partialRight(fn, partialArgs) {
    var leftArgs = [];
    var leftN = fn.length - partialArgs.length;
    while (leftN > 0) {
        leftArgs.push(__);
        leftN = leftN - 1;
    }
    return partial(fn, leftArgs.concat(partialArgs));
}
```

kedashoe replied:
Looks good, I think it is worth it.

kedashoe replied:
:zap:

scott-christopher commented:
Aside from that this is looking great, especially the reduced complexity in the various `_curry*` functions.

scott-christopher commented:
:mushroom: Great work @kedashoe!

(looks like we have a merge conflict that needs to be resolved though)

paldepind commented:
I think  this is wonderful and whould really like to see it merged!

davidchambers commented:
I suggest that we leave this pull request open until we agree on a course of action in #1912. We don't want a repeat of the `reduce` → `foldl` → `reduce` fiasco!

CrossEye commented:
Agreed.
